Improve clarification about communication methods in episode 3 (#50)
Southampton-RSG-Training/dirac-intro-to-mpi: _episodes/03-communicating-data.md

## Communicating data using messages

MPI is a framework for passing data and other messages between independently running processes. If we want
to share or access data from one rank to another, we use the MPI API to transfer data in a "message." A message is a
data structure which contains the data, and is usually expressed as a collection of data elements of a particular data
type.

Sending and receiving data usually happens in two patterns. We either want to send data from one specific rank to
another, known as point-to-point communication, or to/from multiple ranks all at once to a single or multiples targets,
known as collective communication. Whatever we do, we always have to *explicitly* "send" something and to *explicitly*
"receive" something. Data communication can't happen by itself. A rank can't just get data from one rank, and ranks

> don't automatically send and receive data. If we don't program in data communication, data isn'. Unfortunately,

Reviewer: Incomplete sentence, perhaps:

```suggestion
don't automatically send and receive data. If we don't program in data communication, data isn't exchanged. Unfortunately,

```

large amounts data or small amounts too often, those small overheads rapidly add up into a noticeable performance hit.

To get an idea of how communication typically happens, imagine we have two ranks: rank A and rank B. If rank A wants to
send data to rank B (e.g., point-to-point), it must first call the appropriate MPI send function which puts that data
into an internal *buffer*; known as the **send buffer** or **envelope**. Once the data is in the buffer, MPI figures
out how to route the message to rank B (usually over a network) and then sends it to B. To receive the data, rank B must
call a data receiving function which will listen for messages being sent to it. In some cases, rank B will then send
an acknowledgement to say that the transfer has finished, similar to read receipts in e-mails and instant messages.

> ## Check your understanding
>
> Consider a simulation where each rank calculates the physical properties for a subset of cells on a very large
> grid of points. However, the next step of the calculation needs to know the average temperature across the entire grid

> > of points. How would you use calculate the average temperature?

Reviewer: ```suggestion
> of points. How would you calculate the average temperature?

```

>
> > ## Solution
> >
> > There are multiple ways to approach this situation, but the most efficient approach would be to use collective
> > operations to send the average temperature to a main rank which performs the final calculation. You can, of course,
> > also use a point-to-point pattern, but it would be less efficient, especially with a large number of ranks.
> >

> > > If you weren't do this in parallel, or instead using shared-memory parallelism, you wouldn't need to do any

Reviewer: Perhaps add OpenMP as an example of shared memory parallelism, e.g. "...(e.g. using OpenMP)..."?

collection of ranks which are able to exchange data with one another. What this means is that every communication
between two (or more) ranks is linked to a specific communicator in the program. When we run an MPI application, the
ranks will belong to the default communicator known as `MPI_COMM_WORLD`. We've seen this in earlier episodes when, for
example, we've used functions like `MPI_Comm_rank()` to get the rank number,

```c
int my_rank;
MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);  /* MPI_COMM_WORLD is the communicator the rank belongs to */
```

In addition to `MPI_COMM_WORLD`, we can make sub-communicators and distribute ranks into them. Messages can only be
sent and received to and from the same communicator, effectively isolating messages to a communicator. For most
applications, we usually don't need anything other than `MPI_COMM_WORLD`. But organising ranks into communicators can be
helpful in some circumstances, as you can create small "work units" of multiple ranks to dynamically schedule the
workload, or to help compartmentalise the problem into smaller chunks by using a [virtual cartesian

> topology](https://www.mpi-forum.org/docs/mpi-3.1/mpi31-report/node192.htm#Node192). Throughout this lesson, we will

Reviewer: Could we not say "throughout this course" instead? (At least, as long as we don't use any other type of communicator?)

stick to using `MPI_COMM_WORLD`.

## Communication modes

When sending data between ranks, MPI will use one of four communication modes: synchronous, buffered, ready or standard.
When a communication function is called, it takes control of program execution until the send-buffer is safe to be
re-used again. What this means is that it's safe to re-use the memory/variable you passed without affecting the data
that is communicated. If MPI didn't have this concept of safety, then you could quite easily overwrite or destroy any

> data before it is send! This would lead to some very strange behaviour which would be very hard to debug. The difference

Reviewer: ```suggestion
data before it is sent! This would lead to some very strange behaviour which would be very hard to debug. The difference

```
